ssa/manager_apply.go

```diff
@@ -0,0 +1,170 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package ssa
+
+import (
+	"context"
+	"fmt"
+	"sort"
+	"strings"
+	"time"
+
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// Apply performs a server-side apply of the given object if the matching in-cluster object is different or if it doesn't exist.
+// Drift detection is performed by comparing the server-side dry-run result with the existing object.
+// When immutable field changes are detected, the object is recreated if 'force' is set to 'true'.
+func (m *ResourceManager) Apply(ctx context.Context, object *unstructured.Unstructured, force bool) (*ChangeSetEntry, error) {
+	existingObject := object.DeepCopy()
+	_ = m.client.Get(ctx, client.ObjectKeyFromObject(object), existingObject)
+
+	dryRunObject := object.DeepCopy()
+	if err := m.dryRunApply(ctx, dryRunObject); err != nil {
+		if force && strings.Contains(err.Error(), "immutable") {
+			if err := m.client.Delete(ctx, existingObject); err != nil {
+				return nil, fmt.Errorf("%s immutable field detected, failed to delete object, error: %w",
+					FmtUnstructured(dryRunObject), err)
+			}
+			return m.Apply(ctx, object, force)
+		}
+
+		return nil, m.validationError(dryRunObject, err)
+	}
+
+	// do not apply objects that have not drifted to avoid bumping the resource version
+	if !m.hasDrifted(existingObject, dryRunObject) {
+		return m.changeSetEntry(object, UnchangedAction), nil
+	}
+
+	appliedObject := object.DeepCopy()
+	if err := m.apply(ctx, appliedObject); err != nil {
+		return nil, fmt.Errorf("%s apply failed, error: %w", FmtUnstructured(appliedObject), err)
+	}
+
+	if dryRunObject.GetResourceVersion() == "" {
+		return m.changeSetEntry(appliedObject, CreatedAction), nil
+	}
+
+	return m.changeSetEntry(appliedObject, ConfiguredAction), nil
+}
+
+// ApplyAll performs a server-side dry-run of the given objects, and based on the diff result,
+// it applies the objects that are new or modified.
+func (m *ResourceManager) ApplyAll(ctx context.Context, objects []*unstructured.Unstructured, force bool) (*ChangeSet, error) {
+	sort.Sort(SortableUnstructureds(objects))
+	changeSet := NewChangeSet()
+	var toApply []*unstructured.Unstructured
+	for _, object := range objects {
+		existingObject := object.DeepCopy()
+		_ = m.client.Get(ctx, client.ObjectKeyFromObject(object), existingObject)
+
+		dryRunObject := object.DeepCopy()
+		if err := m.dryRunApply(ctx, dryRunObject); err != nil {
+			if force && strings.Contains(err.Error(), "immutable") {
```

Comment:
Given this is used twice, I would factor it out to ensure the logic in both methods continuous to be aligned.